cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+            i -= 1
+            x >>= 1
+        
+        return arr
+        
+    def inner_product(self,x):
+        ''' Returns the amplitude of x'th element of the wavefunction, i.e. <x|psi> '''
+        if type(x) == int:
+            x = self._int2arr(x)
+            
+        mu = sum(x * self.gamma)
+        
+        u = np.zeros(self.n,dtype=bool)
+        for p in range(self.n):
+            if x[p]:
+                u ^= self.F[p,:]
+                mu += 2*(sum(self.M[p,:] & u) % 2)
+                
+        return self.omega*2**(-sum(self.v)/2)*1j**mu*(-1)**sum(self.v & u & self.s)*np.all(self.v | (u==self.s))
+
+    def wave_function(self):
+        wf = np.zeros(2**self.n,dtype=complex)
+        
+        for x in range(2**self.n):
+            wf[x] = self.inner_product(x)
+            
+        return wf
+    
+    def _S(self,q,right=False):
+        if right:
+            self.M[:,q] ^= self.F[:,q]
+            self.gamma[:] = (self.gamma[:] - self.F[:,q]) % 4
+        else:
+            self.M[q,:] ^= self.G[q,:]
+            self.gamma[q] = (self.gamma[q] - 1) % 4
+
+    def _Z(self,q):
+        self._S(q)
+        self._S(q)
+
+    def _X(self,q):
+        self._H(q)
+        self._Z(q)
+        self._H(q)
+
+    def _Y(self,q):
+        self._Z(q)
+        self._X(q)
+        self.omega *= 1j
+            
+    def _CZ(self,q,r,right=False):
+        if right:
+            self.M[:,q] ^= self.F[:,r] 
+            self.M[:,r] ^= self.F[:,q] 
+            self.gamma[:] = (self.gamma[:] + 2*self.F[:,q]*self.F[:,r]) % 4
+        else:
+            self.M[q,:] ^= self.G[r,:] 
+            self.M[r,:] ^= self.G[q,:] 
+            
+    def _CNOT(self,q,r,right=False):
+        if right:
+            self.G[:,q] ^= self.G[:,r] 
+            self.F[:,r] ^= self.F[:,q] 
+            self.M[:,q] ^= self.M[:,r] 
+        else:
+            self.gamma[q] = (self.gamma[q] + self.gamma[r] + 2*(sum(self.M[q,:]&self.F[r,:])%2)) % 4
+            self.G[r,:] ^= self.G[q,:] 
+            self.F[q,:] ^= self.F[r,:] 
+            self.M[q,:] ^= self.M[r,:] 
+            
+    def _H(self,p):
+        t = self.s ^ (self.G[p,:] & self.v)
+        u = self.s ^ (self.F[p,:] & (~self.v)) ^ (self.M[p,:] & self.v)
+        
+        alpha = sum(self.G[p,:] & (~self.v) & self.s) % 2
+        beta = sum(self.M[p,:] & (~self.v) & self.s) 
+        beta += sum(self.F[p,:] & self.v & self.M[p,:])
+        beta += sum(self.F[p,:] & self.v & self.s)
+        beta %= 2
+        
+        delta = (self.gamma[p] + 2*(alpha+beta)) % 4
+        
+        self._update_sum(t,u,delta=delta,alpha=alpha)
+    
+    def _update_sum(self,t,u,delta=0,alpha=0):
+        ''' Implements the transformation (Proposition 4 in Bravyi et al)
+
+                i^alpha U_H (|t> + i^delta |u>) = omega W_C W_H |s'>
+        '''
+        if np.all(t == u):
+            self.s = t
+            self.omega *= 1/np.sqrt(2)*(-1)**alpha*(1+1j**delta)
+        else:
```

Comment:
Add `return` to the if block so the else block can be unindented

Reply:
Done.